### ISSUE TITLE
Change where layouts obtains the title and description for a views block.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#2723]
When updating a site from backdrop 1.6.3 to 1.7.0 I found that some of the descriptions I had previously entered for blocks created within views and appearing on the layouts manage blocks page were no longer editable.

This may be partly due to the fact that some of the blocks concerning me were originally created when the site was developed as a D7 site, and have been imported into Backdrop. 

It seems that in Views it is no longer possible to enter different 'View descriptions' for the subsidiary views within a view, `CHANGE VIEW NAME/DESCRIPTION` will only edit the description of the initial view.

I need to track what is now happening about descriptions in Views and then find out how and where Layouts is finding these descriptions.

~~Revised PR now available at https://github.com/backdrop/backdrop/pull/1909~~

**Comment by Graham-72** [comment]
The issue here is that when creating a custom view block there are six different fields of information about the block, and there is confusion about which of these to use to provide the title and description to be used when the block is used in a layout. A view may comprise more than one display; two of the fields relate to the parent view, four are provided for each child display e.g. a custom view block.

To explain this I have created some screenshots in which I have labelled the fields as follows:
A - the human readable title of the parent view
B - the description of the parent view
1 - a name for each child display
2 - a description for each child display
3 - a title for each child display
4 - a block admin label for each child display

The screenshot below shows these
![home page slide sequence content margaret glass 3](https://cloud.githubusercontent.com/assets/10559274/26771888/012033d2-49ba-11e7-8130-e82467826762.png)


Selecting `CHANGE VIEW NAME/DESCRIPTION` brings up a form to enter values for A and B.
![home page slide sequence content margaret glass](https://cloud.githubusercontent.com/assets/10559274/26771958/781c1b72-49ba-11e7-837b-273d60f103d0.png)

Selecting the Display Name (1) brings up a form to enter value for 1 and 2, note that 2 is not displayed on the main form.
![home page slide sequence content margaret glass 1](https://cloud.githubusercontent.com/assets/10559274/26772176/e14d9b2e-49bb-11e7-8438-bc136395f000.png)


When using the Layouts Manage Blocks display, two of these six fields are used to provide a title and description for the block. At the moment (in release 1.7.0) field 4 is used for the title and field B for the description, producing this result
![home page margaret glass 1](https://cloud.githubusercontent.com/assets/10559274/26772660/c54ec1c0-49be-11e7-84b8-52d2feefb5b6.png)

Note that the code normally provides special arrangements for when a field contains no entry; for this particular demonstration I have entered the text [Block Admin Label] in field 4.

I propose that it would be better to use the title of the display (field 3) and its description (field 2) to provide the title and description of the block, as shown here
![home page margaret glass 2](https://cloud.githubusercontent.com/assets/10559274/26772700/fe217920-49be-11e7-80ea-5b71b30f5c21.png)

If this change is made then we should review the need for the Block Admin Label (field 4) which would then seem to be redundant, and hopefully modify the Views UI so that the display description (field 2) is shown on the main form below the display title (field 3).

**Comment by Graham-72** [comment]
I have now produced a PR https://github.com/backdrop/backdrop/pull/1909 which changes the fields used for block title and description when displayed in Layouts. The aim of the PR is to rationalise the use of the relevant fields and make the result more useful to the user. But it probably does not deal with my original problem of compatibility with previously created views. I may have to edit some of them to conform to the proposed approach.

**Comment by olafgrabienski** [comment]
@Graham-72 I like the idea to rationalise the settings in Views used for block title and description in Layouts. I'd prefer however to have the Views settings for block title and description in Layouts at *one* place. To achieve that, I'd suggest to add a description field to the *Block settings*:

```
Block admin label: 
Block admin description: 
(...)
```
What do you think about it?

**Comment by Graham-72** [comment]
> I'd prefer however to have the Views settings for block title and description in Layouts at one place

@olafgrabienski I very much agree with you about this, and said this in my contribution above

>hopefully modify the Views UI so that the display description (field 2) is shown on the main form below the display title (field 3).

Where we differ is that you are suggesting the we use the field 'Block admin label' as the title of the block (my number 4) rather than the field 'Title' (my number 1), and you may be right. Having both of these fields is confusing to the user and surely we need to eliminate one of them.

Also, we already have a description field (my number 2 in the screenshots above.) If we follow your suggestion this should become your suggested 'Block description field' and placed below the title. I have not yet included this in my PR because I have not yet worked out how to modify the code of views_plugin_display.inc to achieve this.

**Comment by olafgrabienski** [comment]
> Having both of these fields is confusing to the user and surely we need to eliminate one of them.

> Also, we already have a description field (my number 2 in the screenshots above.) If we follow your suggestion this should become your suggested 'Block description field' and placed below the title.

@Graham-72 Thanks for your feedback! I agree that having so many places with information about the view is confusing. I'm however not sure if it's good to eliminate or merge the existing ones. They all make sense holding information for different purposes:

- *View name and description* are displayed in the overview list of all views.
- *Display name* (visible) and *description* (hidden) can hold important information in a view for site builders.
- The *Title* is displayed as title of a display in the frontend (if not overridden, e.g. in the Block configuration).
- *Block admin label* is displayed in Layouts. (In my opinion it makes sense to add a *description* field there because the *Display name* related description field has another purpose.)

**Comment by Graham-72** [comment]
@olafgrabienski thanks for your response. I remain very doubtful about the usefulness of all these fields and am not sure that it is a good idea to add yet another, but I agree that we need to be clear about which title/label and which description appears in Layouts, and then make sure that this is clear in the Views UI. This is the main purpose of raising this issue.

I note also that the Views UI has yet another field for the site builder to use, called 'Comment' under 'Other'.

When you say
>Display name (visible) and description (hidden) can hold important information in a view for site builders.

I am wondering what might be the purpose of this if not visible in Layouts? You say it has another purpose?

**Comment by olafgrabienski** [comment]
> When you say
> 
>> Display name (visible) and description (hidden) can hold important information in a view for site builders.
>
> I am wondering what might be the purpose of this if not visible in Layouts? You say it has another purpose?

As this information is only displayed in the Views UI, I use it differently compared to Layouts related information. For example I tend to use short but more technical *Display names* in Views with many displays (like *Block A* vs. *Block B*) while *Block* as part of the name isn't helpful in a Layout. Then, sometimes I use the *Display description* for reminders regarding the configuration of a views display.

**Comment by Graham-72** [comment]
Thank you for this additional explanation. Because of what you say, I agree it would be good to add an additional field Block Admin Description. But could the description of these fields be more specific than the current wording 'Used in administrative interfaces and not shown to end-users'? Would a mention of Layouts be too specific? Are there other modules that would use these fields?

**Comment by olafgrabienski** [comment]
> But could the description of these fields be more specific than the current wording 'Used in administrative interfaces and not shown to end-users'? Would a mention of Layouts be too specific? Are there other modules that would use these fields?

I agree, a more specific explanation would be helpful. Unfortunately, I have no idea if the Block admin label is meant to be used by other places (or modules).

**Comment by Graham-72** [comment]
In response to the helpful feedback I have revised my PR at backdrop/backdrop#1909 and introduced a new field into the Block Settings of the Views UI. I have named this 'Block layout description' and placed it immediately below the existing 'Block admin label'. The 'Block admin label' is used as the custom views block title in Layouts (this is unchanged) and the new 'Block layout description' is used as the description of the custom block.

**Comment by olafgrabienski** [comment]
I had a look at the changed PR at the sandbox site and get mixed results:

(1) The new 'Block layout description' is used as the description of the custom block. Good!

(2) The 'Block admin label' is NOT used as the custom views block title in Layouts. Instead the following pattern is used: *`View:` View name (Display name)*

(3) When you *add* a block, in the modal window the new 'Block layout description' is used for block title and block description at the same time:

![screen-backdrop-pr-1909-add-block-double-description](https://user-images.githubusercontent.com/14188307/27026699-f3d2323e-4f5e-11e7-982f-91edb57a29c4.png)

**Comment by Graham-72** [comment]
@olafgrabienski thanks for the feedback. I seem to have lost something in creating my PR. I will have another look.

**Comment by Graham-72** [comment]
@olafgrabienski I have made some revisions to my PR. When you have time please can you have another look. A second pair of eyes is a great help, thanks.

**Comment by olafgrabienski** [comment]
You're welcome! I've had another look just now:

(1) The new 'Block layout description' is still used as the description of the custom block. Perfect!

(2) The 'Block admin label' works fine now as views block title in Layouts.

(3) While adding a block, in the modal window the new 'Block layout description' is used correctly, i.e. only for the block description.

So, looks good to me from a site builder perspective! Maybe @klonos can have a look at the code?

**Comment by klonos** [comment]
I am very interested in this issue, but I just finished from a very long day, so taking a look at the PR sandbox made my head spin 😄 ...I'll have to test a bit more and compare the view in the PR sandbox with a duplicate view locally so that I have a better (visual) comparison of before/after. Stay tuned...

**Comment by klonos** [comment]
...my first reaction is: why not use the display name as the block admin label and remove the "Block admin label" setting all together?

**Comment by klonos** [comment]
...also why introduce yet another "Block layout description" setting when we can use the display description?

**Comment by klonos** [comment]
Here is a thing that I personally consider a regression...

Before, the pattern for the title in the autocomplete items was `View: [view-name]: [display-name]`. After, the pattern becomes `View: [block-admin-label]`. So if one tries to search by the view name in the autocomplete...

Before:

![image](https://user-images.githubusercontent.com/2423362/27079049-8718bf20-5079-11e7-9d0b-22cf3ce605ea.png)

After:

![image](https://user-images.githubusercontent.com/2423362/27079036-7c34eaa2-5079-11e7-890b-30973cfe722b.png)

If you must have the "Block admin label" as part of the autocomplete results items, then I would append it like so: `View: [view-name]: [display-name] ([block-admin-label])`.

**Comment by olafgrabienski** [comment]
@klonos Re your questions see https://github.com/backdrop/backdrop-issues/issues/2723#issuecomment-306766033 and the following discussion.

**Comment by Graham-72** [comment]
Maybe it would help if I give my reasoning for wanting to make this change (or something similar). I want to enable a site editor (someone different form me, the site developer) to be able to change at any time the arrangement of some custom blocks on a page, given a fixed layout option of rows and divisions within each row (a customised layout template), and given a set of named custom blocks (many produced by views).

So I need to name each view block display uniquely and give it a meaningful description, bearing in mind that the site editor does not know, or need to know, anything about the view itself. 

My earlier attempt conflicted with a need for the site developer to name and describe each display in a way that met his/her needs, as pointed out by @olafgrabienski above.

So, instead of my original objective of simplifying things and removing one of the fields, we now have an extra field for each block display, two for the specific needs of the site developer, and two for information to the site editor. These are in addition to the two for the 'parent' view.

I think it is very important that somehow we provide a site developer with the tools for making it easy to inform a site editor how to make use of some of the flexibility of Backdrop without expecting them to understand and use the Views UI. In fact, I do not want them to have access to it.

**Comment by Graham-72** [comment]
> Here is a thing that I personally consider a regression...

@klonos please can you double check this because I don't think the PR has caused this, though I may be wrong. What I think you are seeing is the difference between there being an entry in the 'Block admin label' field and there being none.

What I have done in my latest change to the PR is to prefix any entry for Block admin label with 'View: ', which to my mind helps sort the blocks which come from custom views from the rest. I'm not sure about adding in the name of the parent view as well.

I'm now thinking that maybe Block admin label should be a required field.

**Comment by klonos** [comment]
> @klonos please can you double check this because I don't think the PR has caused this, though I may be wrong. What I think you are seeing is the difference between there being an entry in the 'Block admin label' field and there being none.

Double-checked: true about the missing 'Block admin label' setting, but the regression remains even after providing some value for that setting...

Before the PR (vanilla Backdrop 1.7.0):

![image](https://user-images.githubusercontent.com/2423362/27098440-e69c958e-50ba-11e7-9e53-1fa53e0762fb.png)

After the PR:

![image](https://user-images.githubusercontent.com/2423362/27079036-7c34eaa2-5079-11e7-890b-30973cfe722b.png)

**Comment by klonos** [comment]
OK, lets please take baby steps because there is too much inconsistency with what we call things in the UI and what we call the respective variables in the code...

Let's first address how the block "names" (the clickable things) are rendered in the autocomplete list. How about this (`View: [view-name]: [display-name] ([block-admin-label])`):

![image](https://user-images.githubusercontent.com/2423362/27100499-b003ce78-50c1-11e7-98ae-c38e2f947019.png)

...or alternatively this (`[block-admin-label] (View: [view-name]: [display-name])`):

![image](https://user-images.githubusercontent.com/2423362/27100837-c9481e24-50c2-11e7-80b2-8e0dc97ee017.png)

So basically, either append the 'Block admin label' (if it is provided) to how we currently render the name of these items in the autocomplete list, or prepend it.

Lets decide if this is a good step and we'll then move to the description of these items in the list. Once we deal with that, we'll move to the block title and description as they appear when the block is actually placed in a region in the Manage blocks UI. Sounds like a plan?

**Comment by Graham-72** [comment]
@klonos you talk about a regression whent it isn't! You are asking for an additional feature. Fair enough.

I also agree about the confusion in the existing code about the names of fields (in code) not matching their label in some cases.

**Comment by klonos** [comment]
@Graham-72 here's what I've done:

- I have called the view "view-name" on the PR sandbox and on my local vanilla 1.7.0
- I have created 2 blocks ("block-1-name", "block-2-name") and added respective "Block admin label"'s to both ("block-1-admin-label", "block-2-admin-label").
- I have created a 3rd block with no "Block admin label", called it "block-3-no-admin-label".
- Saved the view, headed to a layout and hit the "Add block" button on a region.
- I filtered the list of blocks by the view name ("view-name")

On the vanilla 1.7.0 all 3 blocks show up:

![image](https://user-images.githubusercontent.com/2423362/27105271-4a90ecfe-50d2-11e7-8feb-9b9b6306ac1d.png)

On the PR sandbox, what is returned is the 3rd block only (the one with no "Block admin label" value set):

![image](https://user-images.githubusercontent.com/2423362/27105253-3d3a8772-50d2-11e7-85eb-0de3d0dea04d.png)

When I search by view name, I expect all blocks of that view to be returned (as in the vanilla 1.7.0). Only one is returned. Hence, regression.

The reason for the regression seems to be that with your PR the view name is removed from both the block name and its description (the search filter searches in both).

**Comment by klonos** [comment]
Point is that we need to keep the view name in either the block name or its description. I still insist on taking baby steps:

1. Improve the block titles in the "Add block" list to include the "Block admin label" (if one is provided in the view block display)
2. Improve the block descriptions in the "Add block" list (TBD, but consider  )
3. Decide where the block title is pulled from when the block is added to the UI (TBD, but most likely from the "Block admin label" setting is not empty)
4. Decide where the block description is pulled from when the block is added to the UI (TBD, but most likely either from the block display description or from a new "Block admin description" setting to be added)

5. Clean up variable names along the way so that the code makes sense and try to match with what things are called in the UI (or be as descriptive as possible at least).

I have filed an alternative PR that takes baby step 1 and cleans up the variable names: https://github.com/backdrop/backdrop/pull/1912

**Comment by klonos** [comment]
So from my PR:

![image](https://user-images.githubusercontent.com/2423362/27106607-75b01450-50d7-11e7-8e6b-02955f12016c.png)

@Graham-72 does that make sense? It will be only what happens in the list on the "Add block" dialog and as explained already, I have not changed the descriptions yet (that will be bay step 2).

**Comment by klonos** [comment]
...oh, I just realized what you are trying to achieve @Graham-72. So for you:

"site dev" = somebody that understands views and has permission to create/edit them.
"site-builder" = somebody that does not have access to the views UI, but does have access to the Layouts UI

So, is hiding any views-related info from the block names in the list what you are trying to do?

**Comment by klonos** [comment]
The main reasons why I would like to keep the view-related info (view name, display name) are these:

1. Make sure that the search box in the "Add block" dialog can filter by view name. Perhaps we could add this as a 3rd, hidden field, but...
2. As an advanced user, I still want to be able to tell which view and which block display is the "source" of any "pretty" block name generated by the "Block admin title" setting.

**Comment by Graham-72** [comment]
> I still insist on taking baby steps:

@klonos I am afraid we still do not see eye-to-eye on this and if you are going to 'insist' on your approach then perhaps it would be best if I bow out and leave you to it.

Getting back to why I raised this issue in the first place was that I had tried, when using 1.6.3, to add some 'site editor' friendly descriptions to my custom blocks on one of my sites. I don't think my concept of 'site editor' is quite the same as your 'site builder' because in this case I think of myself as the site developer and builder (same thing?) and someone else as the site editor. Anyway, these descriptions seemed to have disappeared when I updated to 1.7.0. (A regression?). I know that the code has changed but I haven't the energy now to go back and see what exact effect this has had.

I think it is important that the list of blocks on the Layouts page should include the specific name (title) of the display and also a custom description. And that is what I want to achieve. But I do not want to complicate or excessively lengthen the specific name by adding in the view name as well. But I would be happy if this were to be added to the custom description so it reads "[custom description of this display](A block display from the [view name] view)". This would enable your search for the view name to work.

**Comment by Graham-72** [comment]
Here is a screenshot of what we might now show on the Layout Manage blocks page
![screenshot-backdrop margaretglass com-2017-06-14-06-26-21](https://user-images.githubusercontent.com/10559274/27116763-e90f4afe-50ca-11e7-9667-caf1a03b8c2f.png)

This shows three display blocks from the same view. The names and descriptions are examples only. Note that the name is now prefixed with 'View:' and the description suffixed with the phrase 'A block display from the .....'

**Comment by klonos** [comment]
>>I still insist on taking baby steps:
>
>@klonos I am afraid we still do not see eye-to-eye on this and if you are going to 'insist' on your approach then perhaps it would be best if I bow out and leave you to it.

No need for that. I feel like I've hijacked this ticket. Please go on on trying to fix what got broken after 1.6.3. I want to help, if you want me to.

> ...when using 1.6.3, to add some 'site editor' friendly descriptions to my custom blocks on one of my sites. ...these descriptions seemed to have disappeared when I updated to 1.7.0. (A regression?)

If that's what happened, then yes, it is a regression. And we should try to fix that. I can help with testing this. How did you add these site-editor-friendly descriptions? Was it using the views display description (number 2 in your screenshot)?

> I think it is important that the list of blocks on the Layouts page should include the specific name (title) of the display and also a custom description.

Yes. Any reason why the existing view display description (number 2 in your screenshot) will not work as that custom description though? I don't see it being used elsewhere and this use case makes it a perfectly valid candidate. Perhaps we should consider exposing that description instead of hiding it. Perhaps *only for the block views displays*.

> But I do not want to complicate or excessively lengthen the specific name by adding in the view name as well.

Well, with that one I'm not sure I agree. You see, the block names in the "Add block" list use the `View: [view-name]: [display-name]` pattern. If you remove the `[view-name]` part, you have `View: [display-name]`. There's two things wrong with that:

1. It technically is not a view, but a view display. So it should say `View display: [display-name]` or `View block display: [display-name]`. That would be perfectly OK, but...
2. If you have two different views with a block display, you might end up with 2 seemingly identical items in the "Add block" list. Consider this case:

- A "view-name" view with a "block" display.
- Clone the view and call its clone "different-view"

...as it is now, you'd get a "View: view-name: block" block and a "View: different-view: block" block. Easy to tell which is which.

...with your proposed change, you'd get "View: block" and also "View: block". See the problem.

> But I would be happy if this were to be added to the custom description so it reads "[custom description of this display](A block display from the [view name] view)". This would enable your search for the view name to work.

Yes, it would and that's what I now have on my local:

    ![image](https://user-images.githubusercontent.com/2423362/27124574-dcd3cbb0-5134-11e7-923f-b8be17dddb4b.png)

  - When a "Block admin label" is provided in the block display, only that is shown. I don't think that prefixing it with "View: " is accurate. We can add "View display: " or "View block display: ":

    ![image](https://user-images.githubusercontent.com/2423362/27124911-e7eaa27a-5135-11e7-87df-64521569d07e.png)

  - The old, "geeky", non-site-builder-friendly `View: [view-name]: [display-name]` thing is appended in the description (although as a smaller `<pre>`). This way, it's still there for the advanced site dev, while at the same time does not draw attention.

  - When no "Block admin label" is provided, if falls back to the current `View: [view-name]: [display-name]` thing.

  - I have not worked on the descriptions yet on my local, but I agree that we should be able to customize these.

**Comment by klonos** [comment]
@Graham-72 I have left comments in the PR about some coding standard issues + some other suggestions, but my comments seem to be all over the place. I'll try to remove some of them and explain myself here instead.

Some of my suggestions are for pre-existing code and not changes that you have introduced. Nevertheless, I think that we should clean up this function in order to make better sense when people deal with it in the future...

So, throughout the entire `execute_hook_block_list()` function, we seem to be using `$this->view->get_human_name()` and `$this->display->display_title()` multiple times. Better to create two variables for them and use the variables instead.

Also, I think what we are doing here is silly:

```php
if ($this->display->display_title == $this->definition['title']) {
  $block_title = t('View: @view', array('@view' => t($this->view->get_human_name())));
}
else {
  $block_title = t('View: @view: @display', array('@view' => t($this->view->get_human_name()), '@display' => t($this->display->display_title)));
}
```

Better like so:

```php
$block_title = t('View: @view', array('@view' => t($view_name)));
if ($display_title != $this->definition['title']) {
  $block_title.= ': ' . t($display_title);
}
```

**Comment by klonos** [comment]
...now, once we have fixed the silliness with the entire if/else above, lets deal with some other inconsistencies...

1. Lets call the variable that pulls the "Block admin title" setting from the view display `$block_admin_title` in order to match the label of that setting with what we call it in the Views UI:

`$block_admin_title = $this->get_option('block_description');`

2. We need to address the fact that we aim to replace `$block_title` with something else. So this should better be something like `$block_source` (or something like that) instead:

```php
$block_source = t('View: @view', array('@view' => t($view_name)));
if ($display_title != $this->definition['title']) {
  $block_source .= ': ' . t($display_title);
}
```

Now, we can finally do this:

```php
if (empty($block_admin_title)) {
  $block_title = $block_source;
}
else {
  $block_title = t('View block: ') . $block_admin_title;
}
```

Where `$block_admin_title` is the "Block admin title" setting from the view display, `$block_source` is the "old" `View: view-name: display-name` thing and `$block_title` is what we'll be passing into `$delta['info']`:

```php
$delta => array(
  'info' => $block_title,
   ...
```

Does all that make sense @Graham-72?

**Comment by klonos** [comment]
I have updated the issue title to include block title besides description.

**Comment by Graham-72** [comment]
I didn't think we were changing where it gets the block title?

I'm sorry not to have replied yet to many of your comments. I will try and get to them now.

**Comment by Graham-72** [comment]
>Also, I think what we are doing here is silly:

I presume you mean silly in that it is a clumsy way of coding this and your suggestion is better? I think you still have the required three possible outcomes according to what information is available. Perhaps one or two comments in your new code would help the first-time reader.

**Comment by Graham-72** [comment]
> When a "Block admin label" is provided in the block display, only that is shown. I don't think that prefixing it with "View: " is accurate. We can add "View display: " or "View block display: "

I disagree, I think it is accurate because the block derives from a view. When it is listed on the Layouts Manage blocks page it has to be a block, and also a view display, so the extra words are redundant.

Regarding your point about the possibility of two block displays being given the same block admin label, surely it is the responsibility of the site builder to check that all his/her block admin labels are meaningful and distinct. I want to keep what the site editor sees as simple and uncluttered as possible.

**Comment by Graham-72** [comment]
> Any reason why the existing view display description (number 2 in your screenshot) will not work as that custom description though?

I tried this initially but had this response from @olafgrabienski, see a long way above
>Display name (visible) and description (hidden) can hold important information in a view for site builders.

In my view we need provide for a clear description that will be meaningful to a site editor who has the right to use the manage blocks page but does not have access to views. Maybe my aim to allow a site editor to change block layouts with the standard manage blocks UI is too ambitious since perhaps we have a conflict of interest between site developers/builders who understand a lot about the internal workings of Backdrop (e.g. Views) and my site editors who just want to make changes within the framework for their site, working from customised guidance notes provided by their site developer (in this case, me).

**Comment by klonos** [comment]
>>Also, I think what we are doing here is silly:
>
>I presume you mean silly in that it is a clumsy way of coding this and your suggestion is better? I think you still have the required three possible outcomes according to what information is available. Perhaps one or two comments in your new code would help the first-time reader.

Yeah, sorry I wasn't clear. The "silly" code had an if/else statement that was used to set the value of the `$block_title` variable. IF was setting it to "blah blah blah" and ELSE was setting it to the very same "blah blah blah" + "something else". Better to have that variable be "blah blah blah" to begin with, and append the "something else" to it if the opposite of IF happens.

**Comment by klonos** [comment]
>>When a "Block admin label" is provided in the block display, only that is shown. I don't think that prefixing it with "View: " is accurate. We can add "View display: " or "View block display: "

>>I disagree, I think it is accurate because the block derives from a view.

Yeah. We don't seem to see eye to eye on this one. If as a site builder I see "View: Some things", I would just go to the views listing (`/admin/structure/views`) or in the Structure -> Views menu item in the admin menu looking for a view called "Some things". I would not naturally assume that this is a view display name because I saw "View:". Perhaps we should ask for more opinions on this.

**Comment by Graham-72** [comment]
Well, if we go ahead with my suggestion for the extra description field and my more recent addition that we append to it "(A block display from the [view name] view)", the view name would be there, lengthening the description rather than the title field.

**Comment by quicksketch** [comment]
Thanks @Graham-72 for making this issue! I admit the use of front-end and admin labels/descriptions in blocks has been horribly inconsistent.

My thoughts on comments/questions so far:

> But could the description of these fields be more specific than the current wording 'Used in administrative interfaces and not shown to end-users'? Would a mention of Layouts be too specific? Are there other modules that would use these fields?

Yes, we can/should be more specific and explicitly state how this description will be used in Layouts.

I think @olafgrabienski's suggestion that we make an *explicit field* for block description is a good one. That allows the person assembling the view to make a "site-builder friendly" label. I also think that if an explicit label/description is provided, we should use exact that user-provided value in listing and showing blocks.

@klonos's desire to include machine names in labels might be helped if we always make the view machine name searchable. Perhaps we could add it as a "title" attribute and include that in the search?

e.g (proposal):

![selection_205](https://user-images.githubusercontent.com/100206/27254544-81001a48-5358-11e7-9f29-a4116a26efa0.png)

This would also provide an opportunity to do the same thing for custom blocks, for which we also are not displaying the machine names (and I think that's a good thing, for the same reason that @Graham-72 states, he wants to make site-builder friendly names that might not be so developer-sounding).

e.g (proposal):

![image](https://user-images.githubusercontent.com/100206/27254619-db27e2ca-5359-11e7-81a8-0466153f4f43.png)

If we were to make these changes, we might look at updating fields, which currently do display machine names in their labels, but that is because we don't have any concept of block labels or descriptions in Field UI.

e.g (current behavior we may need to revisit):

![image](https://user-images.githubusercontent.com/100206/27254601-7adfaeb6-5359-11e7-8d89-e7cb52e28021.png)

**Comment by Graham-72** [comment]
@quicksketch thank you for your comments on this. I have now modified my PR in order to achieve some of these changes, though more work is needed because tests are failing. I particularly like your suggestion to use the title attribute to provide a tooltip giving the technical information about the view. I have renamed the additional description field to be block_admin_description, and the existing field block_description to be block_admin_title, adding `function views_update_2000()` to views.install in order to modify any existing config records.

I have also used the title attribute for these two block_admin fields to give the tooltip 'Used, for example, in Layouts' in the Views UI in order to give a clearer indication to the user as to the purpose of these two fields.

Still to do:
+ Amend tests to get them working.
+ Make sure the title attribute is included in the search.
+ Apply similar modifications for custom blocks.
+ Look to similarly updating fields.

**Comment by Graham-72** [comment]
>Make sure the title attribute is included in the search.

I do not know how to make this happen. The search would seem to be performed in Javascript in layout.admin.js where we have
   ```
 function showBlockItem(index, row) {
        var $row = $(row);
        var $sources = $row.find('.block-item, .description');
        var textMatch = $sources.text().toLowerCase().indexOf(query) !== -1;
        var $match = $row.closest('div.layout-block-add-row');
        $match.toggle(textMatch);
        if (textMatch) {
          stripeRow($match);
        }
      }
```
The lines of code it is searching are like this
```
<div class="layout-block-add-row form-item form-type-item odd">
 <a class="use-ajax block-item ajax-processed" data-dialog="1"
    href="/admin/structure/layouts/manage/home/add-block/editor/top/views%3Afrontpage_slideshow-block" 
    title="View: Home page slide sequence (frontpage_slideshow)"  
    data-dialog-options="{&quot;dialogClass&quot;:&quot;layout-dialog&quot;}">Slide sequence for home page.</a>
<div class="description">A sequence of images produced by Views Slideshow.</div>
</div>
```
We need the search to include the title attribute. Can anyone help please?

**Comment by Graham-72** [comment]
@quicksketch I also find that the introduction of the title attribute in layout.admin.inc at line 1003 is causing tests to fail. Line 1309 in layout.test is `$this->clickLink(t('Custom block'));` and this fails when the title attribute is added to the link and the link becomes
 `<a href="/admin/structure/layouts/manage/default/add-block/editor/top/layout%3Acustom_block"`
     `title="" `
     `class="use-ajax block-item ajax-processed"` 
     `data-dialog="1" `
     `data-dialog-options="{&quot;dialogClass&quot;:&quot;layout-dialog&quot;}">`
     `Custom block</a>`

Is this a problem with the clickLink test? What else can I try?

**Comment by Graham-72** [comment]
I am re-considering this issue and have deleted the PR and intend to produce a new one, perhaps with a better statement of the issue in the light of all the comments.

**Comment by BWPanda** [comment]
> I am re-considering this issue and have deleted the PR...

Let's close this for now. It's been 3 years an there hasn't been any further discussion or interest here. If @Graham-72 (or anyone else) can re-clarify this problem then maybe open a new issue, considering the length of this one.

**Comment by klonos** [comment]
Because many things have changed between 2017 and now, I have just retested how views, views displays, and views block title descriptions work. I have to admit that it is quite a big mess of UX WTFs as well as bugs. I wanted to raise separate issues, with distinct small scopes, but I frankly do not know where to begin or what titles to give to the issues 😓 😅

**Comment by klonos** [comment]
...actually, with #4558 we may have solved some (most/all?) of the issues here 😅 

Since that change went in, we can provide custom admin labels/descriptions for ALL blocks, including those provided by Views 🎉 

Take the "Promoted content" block that comes with Backdrop core OOTB:

![image](https://user-images.githubusercontent.com/2423362/97118951-c1511280-1715-11eb-8880-5a13ca99dbd6.png)

...configure the block, change the admin label/description:

![image](https://user-images.githubusercontent.com/2423362/97118930-9ff02680-1715-11eb-8f86-cc7104e4ee10.png)

...than save, and ta-da:

![image](https://user-images.githubusercontent.com/2423362/97118941-b26a6000-1715-11eb-9135-fd6c5082285b.png)
